python/cuml/neighbors/nearest_neighbors_mg.pyx

```diff
@@ -0,0 +1,340 @@
+import cuml
+import warnings
+
+from cuml.common.base import Base
+from cuml.utils import get_cudf_column_ptr, get_dev_array_ptr, \
+    input_to_dev_array, zeros, row_matrix
+
+from cython.operator cimport dereference as deref
+
+from cuml.common.handle cimport cumlHandle
+
+import scipy.stats as stats
+
+
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr
+
+import rmm
+from libc.stdlib cimport malloc, free
+from libcpp.vector cimport vector
+
+from libc.stdint cimport uintptr_t, int64_t
+from libc.stdlib cimport calloc, malloc, free
+
+import rmm
+
+
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+
+cdef extern from "cumlprims/opg/matrix/data.hpp" namespace \
+        "MLCommon::Matrix":
+
+    cdef cppclass Data[T]:
+        Data(T *ptr, size_t totalSize)
+
+    cdef cppclass floatData_t:
+        floatData_t(float *ptr, size_t totalSize)
+        float *ptr
+        size_t totalSize
+
+ctypedef Data[int64_t] int64Data_t
+
+
+cdef extern from "cumlprims/opg/matrix/part_descriptor.hpp" namespace \
+        "MLCommon::Matrix":
+
+    cdef cppclass RankSizePair:
+        int rank
+        size_t size
+
+    cdef cppclass PartDescriptor:
+        PartDescriptor(size_t M,
+                       size_t N,
+                       vector[RankSizePair*] &partsToRanks,
+                       int myrank)
+
+cdef extern from "cumlprims/opg/selection/knn.hpp" namespace \
+        "MLCommon::Selection::opg":
+
+    cdef void brute_force_knn(
+        cumlHandle &handle,
+        vector[int64Data_t*] &out_I,
+        vector[floatData_t*] &out_D,
+        vector[floatData_t*] &idx_data,
+        PartDescriptor &idx_desc,
+        vector[floatData_t*] &query_data,
+        PartDescriptor &query_desc,
+        bool rowMajorIndex,
+        bool rowMajorQuery,
+        int k,
+        size_t batch_size,
+        bool verbose
+    ) except +
+
+
+def _free_float_d(data):
+    cdef uintptr_t data_ptr = <size_t>data
+    cdef vector[floatData_t*] *d = <vector[floatData_t*]*>data_ptr
+    for x_i in range(d.size()):
+        free(d.at(x_i))
+    free(d)
+
+
+def _build_float_d(arr_interfaces):
+    """
+    Instantiate a container object for a float data pointer
+    and size.
+    :param arr_interfaces:
+    :return:
+    """
+    cdef vector[floatData_t *] * dataF = new vector[floatData_t *]()
+
+    cdef uintptr_t input_ptr
+    for x_i in range(len(arr_interfaces)):
+        x = arr_interfaces[x_i]
+        input_ptr = x["data"]
+        data = <floatData_t *> malloc(sizeof(floatData_t))
+        data.ptr = < float * > input_ptr
+        data.totalSize = <size_t> (x["shape"][0] *
+                                   x["shape"][1] *
+                                   sizeof(float))
+
+        dataF.push_back(data)
+
+    return < size_t > dataF
+
+
+def _free_mem(index_vec, index_desc,
+              query_vec, query_desc,
+              out_i_vec, out_d_vec,
+              local_index_parts,
+              local_query_parts):
```

Comment:
It might be worth to open a github issue to see if we can generalize a freeing memory method utility function

Reply:
Yeah, I've been thinking about this. As we implement more MG algorithms, we will get a better idea of the reusability of constructs like these. My hope is that we can eventually get to a point where we're passing around a single object to store much of the state. KNN might not be the most reusable example, but I could imagine many of the other algorithms will have a very similar design.

Reply:
#1359